File: chatdownloader/src/leaderboards/nonvips.py
"""
The non-VIPs leaderboard
"""

from typing import Optional

from _types import UserChatPerformance

from .abstractleaderboard import AbstractLeaderboard


class NonVIPS(AbstractLeaderboard):
    """
    Leaderboard for non-VIP users
    """
    @classmethod
    def get_name(cls):
        return 'nonvips'

    def calculate_score(self,
                        performance: UserChatPerformance) -> Optional[float]:
        if 'special_role' not in performance.metadata \
           and not performance.metadata['special_role']:
            return None
        return sum(performance.metrics.values())

File: chatdownloader/src/leaderboards/nonvips.rs
/*
The non-VIPs leaderboard
*/

use crate::leaderboards::leaderboardtrait::AbstractLeaderboard;
use crate::_types::clptypes::UserChatPerformance;
use crate::_types::leaderboardtypes::LeaderboardInnerState;
use std::collections::HashMap;

#[derive(Default, Debug)]
pub struct NonVIPS {
    state: HashMap<String, LeaderboardInnerState>,
}

impl AbstractLeaderboard for NonVIPS {
    fn new() -> Self {
        let mut out = Self {
            state: HashMap::new(),
        };
        out.read_initial_state();
        out
    }

    fn get_name(&self) -> &str {
        "nonvips"
    }

    fn __get_state(&mut self) -> &mut HashMap<String, LeaderboardInnerState> {
        &mut self.state
    }

    fn calculate_score(&self, performance: &UserChatPerformance) -> Option<f32> {
        if performance.metadata.contains_key("special_role") && *performance.metadata.get("special_role").unwrap().get_bool().unwrap() {
            return None;
        }
        return Some(performance.metrics.values().sum());
    }
}